// @aztec/stdlib@0.86.0/dest/kernel/log_hash.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { inspect } from 'util';
import { AztecAddress } from '../aztec-address/index.js';
export class LogHash {
    value;
    counter;
    length;
    constructor(value, counter, length){
        this.value = value;
        this.counter = counter;
        this.length = length;
    }
    toFields() {
        return [
            this.value,
            new Fr(this.counter),
            new Fr(this.length)
        ];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new LogHash(reader.readField(), reader.readU32(), reader.readU32());
    }
    isEmpty() {
        return this.value.isZero() && !this.length && !this.counter;
    }
    static empty() {
        return new LogHash(Fr.zero(), 0, 0);
    }
    toBuffer() {
        return serializeToBuffer(this.value, this.counter, this.length);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new LogHash(Fr.fromBuffer(reader), reader.readNumber(), reader.readNumber());
    }
    toString() {
        return `value=${this.value} counter=${this.counter} length=${this.length}`;
    }
    scope(contractAddress) {
        return new ScopedLogHash(this, contractAddress);
    }
    [inspect.custom]() {
        return `LogHash { ${this.toString()} }`;
    }
}
export class ScopedLogHash {
    logHash;
    contractAddress;
    constructor(logHash, contractAddress){
        this.logHash = logHash;
        this.contractAddress = contractAddress;
    }
    get counter() {
        return this.logHash.counter;
    }
    get value() {
        return this.logHash.value;
    }
    toFields() {
        return [
            ...this.logHash.toFields(),
            this.contractAddress.toField()
        ];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new ScopedLogHash(reader.readObject(LogHash), AztecAddress.fromField(reader.readField()));
    }
    isEmpty() {
        return this.logHash.isEmpty() && this.contractAddress.isZero();
    }
    static empty() {
        return new ScopedLogHash(LogHash.empty(), AztecAddress.ZERO);
    }
    toBuffer() {
        return serializeToBuffer(this.logHash, this.contractAddress);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ScopedLogHash(LogHash.fromBuffer(reader), AztecAddress.fromBuffer(reader));
    }
    toString() {
        return `logHash=${this.logHash} contractAddress=${this.contractAddress}`;
    }
}
