// @aztec/stdlib@0.86.0/dest/validators/schemas.js
import { schemas } from '@aztec/foundation/schemas';
import { z } from 'zod';
export const ValidatorStatusInSlotSchema = z.enum([
    'block-mined',
    'block-proposed',
    'block-missed',
    'attestation-sent',
    'attestation-missed'
]);
export const ValidatorStatusHistorySchema = z.array(z.object({
    slot: schemas.BigInt,
    status: ValidatorStatusInSlotSchema
}));
export const ValidatorStatusHistorySchemaArray = z.array(ValidatorStatusHistorySchema);
export const ValidatorStatusHistorySchemaMap = z.record(ValidatorStatusHistorySchemaArray);
const ValidatorTimeStatSchema = z.object({
    timestamp: schemas.BigInt,
    slot: schemas.BigInt,
    date: z.string()
});
const ValidatorFilteredHistorySchema = z.object({
    currentStreak: schemas.Integer,
    rate: z.number(),
    count: schemas.Integer
});
export const ValidatorStatsSchema = z.object({
    address: schemas.EthAddress,
    lastProposal: ValidatorTimeStatSchema.optional(),
    lastAttestation: ValidatorTimeStatSchema.optional(),
    totalSlots: schemas.Integer,
    missedProposals: ValidatorFilteredHistorySchema,
    missedAttestations: ValidatorFilteredHistorySchema,
    history: ValidatorStatusHistorySchema
});
export const ValidatorsStatsSchema = z.object({
    stats: z.record(ValidatorStatsSchema),
    lastProcessedSlot: schemas.BigInt.optional(),
    initialSlot: schemas.BigInt.optional(),
    slotWindow: schemas.Integer
});
