// @aztec/ethereum@0.86.0/dest/contracts/fee_asset_handler.js
import { EthAddress } from '@aztec/foundation/eth-address';
import { FeeAssetHandlerAbi } from '@aztec/l1-artifacts/FeeAssetHandlerAbi';
import { encodeFunctionData, getContract } from 'viem';
export class FeeAssetHandlerContract {
    txUtils;
    address;
    constructor(address, txUtils){
        this.txUtils = txUtils;
        this.address = EthAddress.fromString(address);
    }
    getMintAmount() {
        const contract = getContract({
            abi: FeeAssetHandlerAbi,
            address: this.address.toString(),
            client: this.txUtils.client
        });
        return contract.read.mintAmount();
    }
    mint(recipient) {
        return this.txUtils.sendAndMonitorTransaction({
            to: this.address.toString(),
            data: encodeFunctionData({
                abi: FeeAssetHandlerAbi,
                functionName: 'mint',
                args: [
                    recipient
                ]
            })
        });
    }
    setMintAmount(amount) {
        return this.txUtils.sendAndMonitorTransaction({
            to: this.address.toString(),
            data: encodeFunctionData({
                abi: FeeAssetHandlerAbi,
                functionName: 'setMintAmount',
                args: [
                    amount
                ]
            })
        });
    }
}
