// @aztec/stdlib@0.86.0/dest/rollup/avm_proof_data.js
import { AVM_PROOF_LENGTH_IN_FIELDS } from '@aztec/constants';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { AvmCircuitPublicInputs } from '../avm/avm_circuit_public_inputs.js';
import { RecursiveProof, makeEmptyRecursiveProof } from '../proofs/recursive_proof.js';
import { VkWitnessData } from '../vks/vk_witness_data.js';
export class AvmProofData {
    publicInputs;
    proof;
    vkData;
    constructor(publicInputs, proof, vkData){
        this.publicInputs = publicInputs;
        this.proof = proof;
        this.vkData = vkData;
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new AvmProofData(reader.readObject(AvmCircuitPublicInputs), RecursiveProof.fromBuffer(reader), reader.readObject(VkWitnessData));
    }
    toBuffer() {
        return serializeToBuffer(this.publicInputs, this.proof, this.vkData);
    }
    static empty() {
        return new AvmProofData(AvmCircuitPublicInputs.empty(), makeEmptyRecursiveProof(AVM_PROOF_LENGTH_IN_FIELDS), VkWitnessData.empty());
    }
}
