// @aztec/stdlib@0.86.0/dest/proofs/proving_request_type.js
export var ProvingRequestType = /*#__PURE__*/ function(ProvingRequestType) {
    ProvingRequestType[ProvingRequestType["PUBLIC_VM"] = 0] = "PUBLIC_VM";
    ProvingRequestType[ProvingRequestType["PRIVATE_BASE_ROLLUP"] = 1] = "PRIVATE_BASE_ROLLUP";
    ProvingRequestType[ProvingRequestType["PUBLIC_BASE_ROLLUP"] = 2] = "PUBLIC_BASE_ROLLUP";
    ProvingRequestType[ProvingRequestType["MERGE_ROLLUP"] = 3] = "MERGE_ROLLUP";
    ProvingRequestType[ProvingRequestType["EMPTY_BLOCK_ROOT_ROLLUP"] = 4] = "EMPTY_BLOCK_ROOT_ROLLUP";
    ProvingRequestType[ProvingRequestType["BLOCK_ROOT_ROLLUP"] = 5] = "BLOCK_ROOT_ROLLUP";
    ProvingRequestType[ProvingRequestType["SINGLE_TX_BLOCK_ROOT_ROLLUP"] = 6] = "SINGLE_TX_BLOCK_ROOT_ROLLUP";
    ProvingRequestType[ProvingRequestType["BLOCK_MERGE_ROLLUP"] = 7] = "BLOCK_MERGE_ROLLUP";
    ProvingRequestType[ProvingRequestType["ROOT_ROLLUP"] = 8] = "ROOT_ROLLUP";
    ProvingRequestType[ProvingRequestType["BASE_PARITY"] = 9] = "BASE_PARITY";
    ProvingRequestType[ProvingRequestType["ROOT_PARITY"] = 10] = "ROOT_PARITY";
    /** Recursive Client IVC verification to connect private to public or rollup */ ProvingRequestType[ProvingRequestType["TUBE_PROOF"] = 11] = "TUBE_PROOF";
    return ProvingRequestType;
}({});
