// @aztec/protocol-contracts@0.86.0/dest/protocol_contract_data.js
// GENERATED FILE - DO NOT EDIT. RUN `yarn generate` or `yarn generate:data`
import { Fr } from '@aztec/foundation/fields';
import { AztecAddress } from '@aztec/stdlib/aztec-address';
export const protocolContractNames = [
    'AuthRegistry',
    'ContractInstanceDeployer',
    'ContractClassRegisterer',
    'MultiCallEntrypoint',
    'FeeJuice',
    'Router'
];
export const ProtocolContractSalt = {
    AuthRegistry: new Fr(1),
    ContractInstanceDeployer: new Fr(1),
    ContractClassRegisterer: new Fr(1),
    MultiCallEntrypoint: new Fr(1),
    FeeJuice: new Fr(1),
    Router: new Fr(1)
};
export const ProtocolContractAddress = {
    AuthRegistry: AztecAddress.fromBigInt(1n),
    ContractInstanceDeployer: AztecAddress.fromBigInt(2n),
    ContractClassRegisterer: AztecAddress.fromBigInt(3n),
    MultiCallEntrypoint: AztecAddress.fromBigInt(4n),
    FeeJuice: AztecAddress.fromBigInt(5n),
    Router: AztecAddress.fromBigInt(6n)
};
export const ProtocolContractLeaves = {
    AuthRegistry: Fr.fromHexString('0x1aa0c7460d3ef24a739cf28e413e712c493d6b37c65f37f9182a7fd924e7e4b7'),
    ContractInstanceDeployer: Fr.fromHexString('0x07ee7acaa2707c287093a1906f7efc139385dfb0e02b39c173ed485bd2844d9a'),
    ContractClassRegisterer: Fr.fromHexString('0x2ce2fd173e93d43d2e82aa3965c41a9ba9364ff2928450191a8a0fa0dc3f7c29'),
    MultiCallEntrypoint: Fr.fromHexString('0x2870569bd41693a61f686e5bff1a4037119c6b9dd016fa276ccc2cb6f148294c'),
    FeeJuice: Fr.fromHexString('0x009196e9a29fc30167c53b9f00f233ecfc0e0fd7a9825a825978ded84270d557'),
    Router: Fr.fromHexString('0x04a9a9bed524c29ed1a52a603fdb2f3c5e5e5941b7ae81fc39b13d93701ec4c4')
};
export const protocolContractTreeRoot = Fr.fromHexString('0x1ed87d453cb7c21c5dbca65441925b3f9138fd1433928dda23342367bf94f7c0');
export const REGISTERER_CONTRACT_CLASS_REGISTERED_TAG = Fr.fromHexString('0x20cd27645f65c15a38720ca984aef936592f3938a705db706d068ad13fc58ae9');
export const REGISTERER_PRIVATE_FUNCTION_BROADCASTED_TAG = Fr.fromHexString('0x0abf338154e355442d9916d630da0491036d5a7b23c42b477c1395ba3bf9b42d');
export const REGISTERER_UTILITY_FUNCTION_BROADCASTED_TAG = Fr.fromHexString('0x27774539896271d6dbe0c1de45ddc44709d1e8d9a9fb686ca3c105d87c417bb8');
export const DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_TAG = Fr.fromHexString('0x2ec28b91a5f838506d6042915005ff55cf7a0a5f889a83b11faed33a31b486f2');
export const DEPLOYER_CONTRACT_INSTANCE_UPDATED_TAG = new Fr(1534834688047131268740281708431107902615560100979874281215533519862n);
